File: raffia-core/src/main/java/com/codepoetics/raffia/paths/segments/BasePathSegment.java
package com.codepoetics.raffia.paths.segments;

import com.codepoetics.raffia.baskets.Basket;
import com.codepoetics.raffia.operations.Projector;
import com.codepoetics.raffia.operations.Projectors;
import com.codepoetics.raffia.operations.Updater;
import com.codepoetics.raffia.paths.Path;
import com.codepoetics.raffia.paths.PathSegment;

abstract class BasePathSegment implements PathSegment {

  @Override
  public Updater createUpdater(Path path, Updater updater) {
    final Path subPath = path.tail();
    final Updater continuation = subPath.isEmpty()
        ? updater
        : subPath.head().createUpdater(subPath, updater);

    return createUpdater(continuation);
  }

  @Override
  public Updater createItemUpdater(Path tail, Updater updater) {
    throw new UnsupportedOperationException("Cannot create item updater for non-conditional path segment");
  }

  @Override
  public Projector<Basket> createItemProjector(Path tail) {
    throw new UnsupportedOperationException("Cannot create item projector for non-conditional path segment");
  }

  protected abstract Updater createUpdater(Updater continuation);

  @Override
  public Projector<Basket> createProjector(Path path) {
    final Path subPath = path.tail();
    return subPath.isEmpty()
        ? createProjector(Projectors.id())
        : createProjector(subPath.head().createProjector(subPath));
  }

  @Override
  public boolean isConditional() {
    return false;
  }

  protected abstract Projector<Basket> createProjector(Projector<Basket> continuation);

}

File: raffia-core/src/main/java/com/codepoetics/raffia/paths/segments/BasePathSegment.kt
package com.codepoetics.raffia.paths.segments

import com.codepoetics.raffia.baskets.Basket
import com.codepoetics.raffia.functions.Projector
import com.codepoetics.raffia.operations.Projectors
import com.codepoetics.raffia.functions.Updater
import com.codepoetics.raffia.paths.Path
import com.codepoetics.raffia.paths.PathSegment

internal abstract class BasePathSegment : PathSegment {

    override fun createUpdater(path: Path, updater: Updater): Updater {
        val subPath = path.tail()
        val continuation = if (subPath.isEmpty)
            updater
        else
            subPath.head().createUpdater(subPath, updater)

        return createUpdater(continuation)
    }

    override fun createItemUpdater(tail: Path, updater: Updater): Updater {
        throw UnsupportedOperationException("Cannot create item updater for non-conditional path segment")
    }

    override fun createItemProjector(tail: Path): Projector<Basket> {
        throw UnsupportedOperationException("Cannot create item projector for non-conditional path segment")
    }

    protected abstract fun createUpdater(continuation: Updater): Updater

    override fun createProjector(path: Path): Projector<Basket> {
        val subPath = path.tail()
        return if (subPath.isEmpty)
            createProjector(Projectors.id)
        else
            createProjector(subPath.head().createProjector(subPath))
    }

    override val isConditional: Boolean
        get() = false

    protected abstract fun createProjector(continuation: Projector<Basket>): Projector<Basket>

}
